subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildServiceIntegrationTest.groovy

```diff
@@ -288,4 +289,52 @@ class ConfigurationCacheBuildServiceIntegrationTest extends AbstractConfiguratio
         outputContains 'probe(classloader2) => 5'
         outputContains 'probe(classloader2) => 6'
     }
+
+    @ToBeImplemented("https://github.com/gradle/gradle/issues/22337")
+    def "build service can be used as input of value source obtained at configuration time"() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+        buildFile("""
+            abstract class EmptyService implements BuildService<${BuildServiceParameters.name}.None> {
+                int getValue() {
+                    return 42
+                }
+            }
+
+            abstract class ServiceValueSource implements ValueSource<Integer, Params> {
+                interface Params extends ValueSourceParameters {
+                    Property<EmptyService> getService()
+                }
+                @Override
+                Integer obtain(){
+                    return parameters.service.get().value
+                }
+            }
+            def serviceProvider = gradle.sharedServices.registerIfAbsent("counter", EmptyService) {}
+            def valueSource = providers.of(ServiceValueSource) {
+                parameters {
+                    service = serviceProvider
+                }
+            }.get()
+
+            task check {
+                doLast {
+                    println "valueSource = " + valueSource
+                }
+            }
+        """)
+
+        when:
+        configurationCacheRun "check"
+
+        then:
+        outputContains("valueSource = 42")
+        configurationCache.assertStateStored()
+
+        // TODO(https://github.com/gradle/gradle/issues/22337) Either the load should succeed, or the clear error message should be provided
```

Comment:
🟡 I'd expect a clear error message / problem report at store time instead of a successful run since the infrastructure to instantiate build services is not available at the time we validate the fingerprint of value sources used at configuration time.

Reply:
I've updated the TODO here to match your suggestion